aten/src/ATen/native/quantized/cudnn/Pooling.cpp

```diff
@@ -0,0 +1,208 @@
+#ifdef USE_CUDA
+#include <ATen/cuda/CUDAConfig.h>  // for the definition of AT_CUDNN_ENABLED
+
+#if AT_CUDNN_ENABLED()
+
+#include <ATen/native/cudnn/Macros.h>
+
+#if HAS_CUDNN_V8()
+
+#include <ATen/ATen.h>
+#include <ATen/cuda/Exceptions.h>
+#include <ATen/cudnn/Descriptors.h>
+#include <ATen/cudnn/Handle.h>
+#include <ATen/cudnn/Types.h>
+#include <ATen/native/Pool.h>
+#include <ATen/native/TensorIterator.h>
+#include <c10/core/ScalarType.h>
+#include <c10/util/ArrayRef.h>
+#include <torch/library.h>
+
+#include <vector>
+
+namespace at {
+namespace native {
+namespace {
+// TODO: This function is the same as that of qpool.cpp. We should refactor this into quantized directory
+// so that we don't need to duplicate the function
+void check_maxpool2d_params(
+    IntArrayRef kernel_size,
+    IntArrayRef stride,
+    IntArrayRef padding,
+    IntArrayRef dilation) {
+  TORCH_CHECK(kernel_size.size() == 1 || kernel_size.size() == 2,
+              "Expected 1d or 2d kernel size, got ", kernel_size.size());
+  TORCH_CHECK(stride.empty() || stride.size() == 2,
+              "Expected no strides or 2d strides, got", stride.size());
+  TORCH_CHECK(padding.size() == 1 || padding.size() == 2,
+              "Expected 1d or 2d padding, got ", padding.size());
+  TORCH_CHECK(dilation.size() == 1 || dilation.size() == 2,
+              "Expected 1d or 2d dilation, got ", dilation.size());
+}
+}
+
+// Currently we support 4D and 3D input (qx) tensors, the latter of which is supported for
+// legacy reasons. The first dimension of a 4D input tensor is the batch size.
+// For a 3D tensor, there is no batch size dimension -- it can be viewed as a single batch.
+// cudnn's 2D pooling operation requires the input and output to be 4D tensors, so we must cast
+// any 3D tensors to 4D prior to using cudnn
+// This implementation currently uses the v7 cudnn APIs as v8 cudnn APIs are not yet available for
+// pooling operations.
+// Consult https://docs.nvidia.com/deeplearning/cudnn/api/index.html#cudnnPoolingForward for
+// documentation on the APIs
+// Currently, it appears there is no cudnn support for dilated pooling -- we will
+// submit a feature request for this with cudnn
+Tensor quantized_max_pool2d_cudnn(
+    const Tensor& qx,
+    IntArrayRef kernel_size,
+    IntArrayRef stride,
+    IntArrayRef padding,
+    IntArrayRef dilation,
+    bool ceil_mode) {
+  check_maxpool2d_params(
```

Comment:
so the dimension checks etc. can be removed after we have structured kernel support right? can we list the blockers for using that and add a TODO so that we can fix this later?

Reply:
I think the structured kernel support for dimension checking is already there (just like it was there for the quantized cpu max pool 2d), but the blocker here is cudnn doesn't allow us to obtain indices for max pool2d.
so I don't think there's a way to implement max_pool2d_with_indices_out_quantized_cuda unless we want to return nothing for indices?

Reply:
oh I see, maybe we can add this context in the comment